umi_tools/sam_methods.py

```diff
@@ -411,10 +411,18 @@ def __call__(self, inreads):
                 try:
                     umi, cell = self.barcode_getter(read)
                 except KeyError:
-                    error_msg = "Read skipped, missing umi and/or cell tag"
+                    error_msg = "Read skipped, missing umi and/or cell tag"  
                     if self.read_events[error_msg] == 0:
+                        
+                        # pysam renamed .tostring -> to_string in 0.14
+                        # .tostring requies access to the parent AlignmentFile
+                        try:
+                            formatted_read = read.to_string()
+                        except AttributeError:
+                            formatted_read = read.query_name
```

Comment:
This is just the name of the read though, right? Ideally  `formatted_read` is the full BAM entry as a string. Does `str(read)` for all pysam versions by any chance, e.g does pysam set the `.__str__()` method?

Reply:
If it's not possible to return the complete entry as a string in all versions of pysam, happy for pysam<0.14 to just return the read name

Reply:
In pysam < 0.14, the read object doesn't contain all the information neccessary to convert the read to text.